python/ql/lib/semmle/python/ApiGraphs.qll

```diff
@@ -474,6 +504,73 @@ module API {
     int getNumArgument() { result = count(this.getArg(_)) }
   }
 
+  /**
+   * A read of a subscript that is connected to the API graph.
+   *
+   * This can be used to reason about the index of subscripts.
+   * For the code
+   * ```python
+   * x = foo[bar]
+   * y = foo[baz]
+   * ```
+   * there wil be subscript edges from `foo` to both `foo[bar]` and `foo[baz]`,
+   * and there will be index edges from `foo` to both `bar` and `baz`.
+   * There will be a separate `SubscriptReadNode` for `foo[bar]` (and one for `foo[baz]`),
+   * and the member predicates `getObject` and `getIndex` will return nodes
+   * consistent with that particular read.
+   */
+  class SubscriptReadNode extends DataFlow::Node {
+    API::Node subscripted;
+
+    SubscriptReadNode() { this = subscripted.getASubscript().asSource() }
+
+    /** Gets an API node for the object being subscripted. */
+    API::Node getObject() { result = subscripted }
+
+    /** Gets an API node representing the index of this read. */
+    API::Node getIndex() {
+      result = subscripted.getIndex() and
+      result.asSink().asCfgNode() = this.asCfgNode().(PY::SubscriptNode).getIndex()
+    }
+  }
+
+  /**
+   * A write to a subscript that is connected to the API graph.
+   *
+   * This can be used to reason about the index and written values of subscripts.
+   * For the code
+   * ```python
+   * foo[bar] = x
+   * foo[baz] = y
+   * ```
+   * there wil be subscript edges from `foo` to both `x` and `y`,
+   * and there will be index edges from `foo` to both `bar` and `baz`.
+   * There will be a separate `SubscriptWriteNode` for `x` (and one for `y`),
+   * and the member predicates `getObject`, `getIndex`, and `getValue` will return nodes
+   * consistent with that particular write.
+   */
+  class SubscriptWriteNode extends DataFlow::Node {
+    API::Node subscripted;
+
+    SubscriptWriteNode() { this = subscripted.getASubscript().asSink() }
+
+    /** Gets an API node for the object being subscripted. */
+    API::Node getObject() { result = subscripted }
+
+    /** Gets an API node representing the index of this write. */
+    API::Node getIndex() {
+      result = subscripted.getIndex() and
+      exists(PY::SubscriptNode subscriptNode |
+        subscriptNode.(PY::DefinitionNode).getValue() = this.asCfgNode() and
```

Comment:
Indeed, the idea is that other syntax could also constitute a subscript access. In the example you give, the read and the write will have different indices.

Reply:
My point wasn't that it should only cover the `a[b] = c` syntax, but rather that it doesn't make much sense to me to tie it to `c`.

Perhaps a comparison with [`AttrWrite`](https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/dataflow/new/internal/Attributes.qll#L79) is useful. Here, for the statement `a.b = c`, the `AttrWrite` node is associated with `a.b`, not `c`. Of course, `c` is still available via `getValue`.

Similarly, if and when we add support for interpreting `a.__setitem__(b, c)` as a subscript write, I would expect the entire expression to be the `SubscriptWriteNode`, not just the `c` part.

Reply:
The underlying `DefinitionNode` is associated with the _target_ of the assignment, not the value or the whole assignment: https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/Flow.qll#L625

(Indeed, it must, otherwise your inline cast would fail.)

To put a finer point on it, say I want to _only_ select those `SubscriptWriteNode`s that correspond to the form `a[b]`, and not the `__setitem__` form. It would be very intuitive for me to write, say, `this.asCfgNode() instanceof SubscriptNode`, but that wouldn't do what I expected!